### ISSUE TITLE
Prioritize/rank RS search results by relevance (learning center)

### ISSUE THREAD

**Issue by jenniferlee-dsva** [department-of-veterans-affairs/va.gov-team#16915]
## Prioritize/rank RS search results by relevance (learning center)

**User story:** 
As a Veteran searching for info in the RS, when I get a list of  results, I want to see the most related/relevant articles first. 

Currently, the search does not rank or prioritize results. We  simply list results in alpha order by title. While this works ok when there are only a few results, it will be unhelpful when there are several results or more than one page.

**Goal: We want to add logic that will rank or prioritize results by relevance.** 

- One way we might sniff relevance is to rank by how many of the relevant words in the search query appear in the `article title `and `body text.` Per grooming discussion, we'll start this by adding 3 sorting logic to prioritize (list first) most relevant results.

**AC:**

- [ ] 1. Sort first by query word instances found in title descending
- [ ] 2. Sort ties then by query word instances found in body descending. Limit to article body text, not text that appears in global layers like the `related info`, `need more help`, `benefit hubs`, and other bread layer sections.
- [ ] 3. Sort ties then by alphabetical descending
- [ ] 4. Hide new sorting logic on production - test and validate in staging/dev or demo environment

So using the example in the screenshot below, the 2 change address articles should appear at the top before the direct deposit step-by-steps (which just happens to have the word "change" in the title). 

See example screenshot below. In this instance, we searched on "change of address." The articles that have both "change" and "address" should be listed first, before the articles with just one of the search words.   _Note: We'll want to include phrase matching for the most relevance sniffing, but we will do that as a separate ticket._ 

![rank-results-by-relevance-example-120320.png](https://images.zenhubusercontent.com/5bb2975c143b4d2ee7a35ef6/7fc66e05-e8c6-4f38-8771-7ac3ee7230c5)

**Comment by brianalloyd** [comment]
Needs to be tested in Dev/Staging environment first, per Grooming session.

**Comment by kelsonic** [comment]
To be clear, could we update the AC to include:

1. Sort first by query word instances found in title descending
2. Sort ties then by query word instances found in body descending
3. Sort ties then by alphabetical descending
4. Hide new sorting logic on production

CC @jenniferlee-dsva @youngfreezyVA @brianalloyd

**Comment by jenniferlee-dsva** [comment]
updated AC with grooming details, per Kelson's comment ^

**Comment by kelsonic** [comment]
If this one isn't being worked on, I'd love to take this for next sprint! Seems super fun 🙂

**Comment by jenniferlee-dsva** [comment]
That would be awesome @kelsonic . However, only if we can confidently deliver [#15649](https://app.zenhub.com/workspaces/vft-59c95ae5fda7577a9b3184f8/issues/department-of-veterans-affairs/va.gov-team/15649#issuecomment-743323652) before code freeze. I want to make sure the content team is unblocked from RS content publishing before our last code merge.

**Comment by kelsonic** [comment]
PR is up for this 🙂 should be merged by EOD, able to validate on Monday

**Comment by jenniferlee-dsva** [comment]
Thank you @kelsonic - will you add here or slack me the staging link where I can validate it?

**Comment by brianalloyd** [comment]
Thanks @kelsonic looking forward to reviewing.

**Comment by jenniferlee-dsva** [comment]
### After sorting update:

**For "change of address"**
![image.png](https://images.zenhubusercontent.com/5bb2975c143b4d2ee7a35ef6/805e6395-8424-4760-9a6b-38c070a8a04e)

**For "claim status" - the numbers = full intro text count of query instances**
![image.png](https://images.zenhubusercontent.com/5bb2975c143b4d2ee7a35ef6/c0c3ca68-6d9c-454b-b831-f1db7d60ce35)

**Comment by jenniferlee-dsva** [comment]
Also checked "discharge status." Here are the results differences btn prod (non-sort/ranked logic) and with sort/ranking logic. 

Non sort/ranked:

**Comment by jenniferlee-dsva** [comment]
Other queries I checked against prod - the sort logic didn't show any effect on these queries. 

`"can't log in"` <-- this was interesting. both prod and sorted had same results in same order, but both were missing all of the signing in articles bc our sign in articles don't use the words "log in." I think they would have come up if we included searching tags. (We have a tag called "Sign in.") So that might be additional info scent to include for a different PR.

`"sign in"`  <-- only 1 article. this one is interesting too bc there are a few articles that include "sign in" in the article text, but only 1 with it in the title. so the logic seems to require that the user's query term is in the title. And if the query is in the article but not in the title, it's not included in results.  I think this is okay for now. We'll have to see how this plays out when there is more content and more searches in RS. 

`"vic card"` <- only 1 article with VIC in title and article body

The sort/ranking logic works as defined by the AC.  I think this is okay for now. We'll have to see how this plays out when there is more content and more searches in RS.  Validated and closing.

**Comment by jenniferlee-dsva** [comment]
Adding Kelson's Slack explanation on how the logic in this ticket is implemented.
[Reference Slack thread](https://dsva.slack.com/archives/C52CL1PKQ/p1609428369068400?thread_ts=1609428198.067300&cid=C52CL1PKQ)

>currently the filtering and sorting logic is separated, the work I did was purely sorting:
current filtering logic is:
Lowercase the article's title and split up into a list of words
Check to see if any of the search term words match the beginning of any of the articles' title's words

>current sorting logic is:
Sort first by query word instances found in title descending
Sort ties then by query word instances found in description descending
Sort ties then by alphabetical descending
